### PR TITLE
core/vm, core/state: return error if gas is insufficient for cost

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

File: core/state/access_events.go [hunk 1]
```diff
@@ -91,67 +91,104 @@ func (ae *AccessEvents) Copy() *AccessEvents {
...
+// available gas is insufficient to cover the total cost, a false flag is
+// returned.
+func (ae *AccessEvents) ValueTransferGas(callerAddr, targetAddr common.Address, availableGas uint64) (uint64, bool) {
+	var gas uint64
+	cost, sufficient := ae.touchAddressAndChargeGas(callerAddr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas)
+	if !sufficient {
+		return 0, false
 	}
-	_, expected2 := ae.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas-expected1)
-	if expected1+expected2 > availableGas {
```
Comment (rjl493456442) [inline core/state/access_events.go:130]: Shouldn't it be `expected1 + expected2`?

File: core/state/access_events.go [hunk 2]
```diff
@@ -91,67 +91,104 @@ func (ae *AccessEvents) Copy() *AccessEvents {
...
+
+	return gas, true
 }
 
-// ContractCreatePreCheckGas charges access costs before
-// a contract creation is initiated. It is just reads, because the
-// address collision is done before the transfer, and so no write
-// are guaranteed to happen at this point.
-func (ae *AccessEvents) ContractCreatePreCheckGas(addr common.Address, availableGas uint64) uint64 {
-	consumed, expected1 := ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BasicDataLeafKey, false, availableGas)
```
Comment (rjl493456442) [inline core/state/access_events.go:141]: if consumed < expected1, then the `availableGas-consumed` will be incorrect

File: core/vm/eips.go [hunk 3]
```diff
@@ -356,15 +356,18 @@ func opPush1EIP4762(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext
...
 
+		// The PC is not exposed in the gas func.
+		// TODO(rjl493456442) we either need to refactor the gas func definition
+		// a bit, or keep the hack here...
 		if !scope.Contract.IsDeployment && !scope.Contract.IsSystemCall && *pc%31 == 0 {
 			// touch next chunk if PUSH1 is at the boundary. if so, *pc has
 			// advanced past this boundary.
 			contractAddr := scope.Contract.Address()
-			consumed, wanted := interpreter.evm.AccessEvents.CodeChunksRangeGas(contractAddr, *pc+1, uint64(1), uint64(len(scope.Contract.Code)), false, scope.Contract.Gas)
-			scope.Contract.UseGas(wanted, interpreter.evm.Config.Tracer, tracing.GasChangeUnspecified)
```
Comment (rjl493456442) [inline core/vm/eips.go:364]: It's very weird.

Theoretically, if the available gas is not sufficient, we should terminate the execution and deduct all the remaining gas outside.

It's very weird to deduct the gas cost regardless of the available gas is sufficient or not...

File: core/vm/evm.go [hunk 4]
```diff
@@ -541,13 +541,12 @@ func (evm *EVM) initNewContract(contract *Contract, address common.Address) ([]b
 			return ret, ErrCodeStoreOutOfGas
 		}
 	} else {
-		consumed, wanted := evm.AccessEvents.CodeChunksRangeGas(address, 0, uint64(len(ret)), uint64(len(ret)), true, contract.Gas)
-		contract.UseGas(consumed, evm.Config.Tracer, tracing.GasChangeWitnessCodeChunk)
-		if len(ret) > 0 && (consumed < wanted) {
```
Comment (rjl493456442) [inline core/vm/evm.go:546]: the `len(ret) > 0` is unnecessary, as the gas cost for empty code is always zero.

File: core/vm/operations_verkle.go [hunk 5]
```diff
@@ -209,8 +236,11 @@ func gasExtCodeCopyEIP4762(evm *EVM, contract *Contract, stack *Stack, mem *Memo
 	}
 	code := evm.StateDB.GetCode(addr)
 	_, copyOffset, nonPaddedCopyLength := getDataAndAdjustedBounds(code, uint64CodeOffset, length.Uint64())
-	_, wanted := evm.AccessEvents.CodeChunksRangeGas(addr, copyOffset, nonPaddedCopyLength, uint64(len(code)), false, gas)
```
Comment (rjl493456442) [inline core/vm/operations_verkle.go:212]: we should use contract.Gas - gas

**Review by gballet — COMMENTED** [review]
It makes sense and it seems to work with the execution spec tests. I am quite surprised that the contract creation passes, I would have to check with the testnet

File: core/vm/evm.go (hunk 4 above)
Comment (gballet) [reply core/vm/evm.go:546]: this seems to come from a rebase, I agree this is now useless

**Comment by gballet** [comment]
> (a) During gas cost computation, the access list may be mutated. Some operations may require access to multiple state entries. It’s possible that part of the access succeeds and gets cached in the access list, while the remaining accesses fail due to an out-of-gas (OOG) error.
> 
> The semantics around whether the successfully cached entries should be reverted in such cases is unclear. Under the current implementation, these entries are not reverted, which could lead to inconsistent behavior in subsequent gas cost computations.

That's not an issue: we are using this to build the witness, and the gas costs correspond to that use case. Once the witness has been paid for, the next access in the same transaction will be cheaper, we want it to be this way.

If the transaction reverts, then the next transaction will use a new access list, and so the witness will be cleared for future gas costs.

> 
> (b) Technically, it’s unnecessary to compute the exact gas cost if the available gas is already insufficient. A simpler alternative is to have the gas cost function return an error when there’s not enough gas. This error can then be propagated up the call stack, ultimately terminating EVM execution and consuming all remaining gas.
> 
> For reference, see the implementation in gasCallEIP2929: https://github.com/ethereum/go-ethereum/blob/master/core/vm/operations_acl.go#L196
> 
> In the current implementation, the required gas cost is also ambiguous. e.g. if two state entries are accessed in the operation, the real gas cost should be cost_1 + cost_2, unfortunately in some places, only cost_1 is returned. I would prefer to simply return an error.

That is indeed a very nice simplification, I am not very happy with the fact that "we don't care about the exact amount of gas being consumed". I  agree that it doesn't seem to have any impact and so we don't really care since everything will be consumed. We just need to make sure that it won't create any corner case. I just want to test it very thoroughly before I give it a go.

**Comment by gballet** [comment]
as it were, I discovered that the execution spec tests were just skipping all the tests while saying 'passed'. Among other issues, it means that moving op -> gas broke but wasn't reported as broken and I have to run through a few hoops to see if this PR really works (I mean, if it's more broken than the base branch is)

**Comment by rjl493456442** [comment]
> That's not an issue: we are using this to build the witness, and the gas costs correspond to that use case. Once the witness has been paid for, the next access in the same transaction will be cheaper, we want it to be this way.

Let's say we want to access account basic data and account code hash in a single operation, and gas is only sufficient for the step 1. At that point the account basic data hasn't been resolved yet and shouldn't be included in the witness.

> If the transaction reverts, then the next transaction will use a new access list, and so the witness will be cleared for future gas costs.

This can also occur in deeper call frames. For example, in opCall, several state entries may be cached in the access event before the operation reverts due to an out-of-gas error. In such cases, should we clear the cached entries? Or should we leave them as-is—knowing they can affect gas cost computation in subsequent operations within the same transaction, but from a different call frame?

File: core/state/access_events.go (hunk 1 above)
Comment (rjl493456442) [reply core/state/access_events.go:130]: in your implementation, the required gas is `expected1+expected2`, but the returned value is `params.WarmStorageReadCostEIP2929` which i believe is a mistake

File: core/state/access_events.go (hunk 2 above)
Comment (rjl493456442) [reply core/state/access_events.go:141]: i would prefer to using my approach to aviod the complexity in the first place

**Comment by gballet** [comment]
> > That's not an issue: we are using this to build the witness, and the gas costs correspond to that use case. Once the witness has been paid for, the next access in the same transaction will be cheaper, we want it to be this way.
> 
> Let's say we want to access account basic data and account code hash in a single operation, and gas is only sufficient for the step 1. At that point the account basic data hasn't been resolved yet and shouldn't be included in the witness.

it should be, because otherwise the stateless client will not be able to read back the basic data when trying to verify the block. Now, I know you want to do everything atomic, but that's not the spec and the spec can change but it takes time and we're in agreement that the first step is to sync the verkle testnet. This is the behavior that is needed to sync the verkle testnet.

> 
> > If the transaction reverts, then the next transaction will use a new access list, and so the witness will be cleared for future gas costs.
> 
> This can also occur in deeper call frames. For example, in opCall, several state entries may be cached in the access event before the operation reverts due to an out-of-gas error. In such cases, should we clear the cached entries? 

no

> Or should we leave them as-is—knowing they can affect gas cost computation in subsequent operations within the same transaction, but from a different call frame?

yes, we should do that. The gas costs are about paying for increasing the witness size. If an addition to the witness was paid for by a transaction, it's fair that any subsequent accesses for that transaction will not pay the full cost.

File: core/state/access_events.go (hunk 1 above)
Comment (gballet) [reply core/state/access_events.go:130]: right it's a rebase error, in the [source branch](https://github.com/gballet/go-ethereum/blob/4a2f6fbaef0025260932c1ef95182090ffdce871/core/state/access_witness.go#L120) it's the correct behavior

File: core/state/access_events.go [hunk 6]
```diff
@@ -91,67 +91,104 @@ func (ae *AccessEvents) Copy() *AccessEvents {
...
+
+	return gas, false
 }
 
 // ContractCreateInitGas returns the access gas costs for the initialization of
-// a contract creation.
-func (ae *AccessEvents) ContractCreateInitGas(addr common.Address, availableGas uint64) (uint64, uint64) {
+// a contract creation. If the available gas is insufficient to cover the total
+// cost, a false flag is returned.
+func (ae *AccessEvents) ContractCreateInitGas(addr common.Address, availableGas uint64) (uint64, bool) {
```
Comment (gballet) [inline core/state/access_events.go:177]: This single change breaks 243 execution spec tests out of 532
